File: Learn/DataStructure/UnionFindSet/template.cpp
class UnionFindSet{
	private:
		int setSize = 0;
		unordered_map<int, int> father;
	public:
		void add(int x) {
			if (!this->father.count(x)) {
				this->father[x] = -1;
				this->setSize++;
			}
		}
		int find(int x) {
			int root = x;
			while (this->father[root] != -1) {
				root = this->father[root];
			}
			while (x != root) {
				int o = this->father[x];
				this->father[x] = root;
				x = o;
			}
			return root;
		}
		void merge(int x, int y) {
			int rootX = this->find(x);
			int rootY = this->find(y);
			if (rootX != rootY) {
				this->father[rootX] = rootY;
				this->setSize--;
			}
		}
		bool isUnion(int x, int y) {
			return this->find(x) == this->find(y);
		}
}; 

File: 算法与数据结构之美/DataStructure/UnionFindSet/template.py
class UnionFindSet:
    def __init__(self, n: int):
        self.count = n
        self.father, self.size = [i for i in range(n)], [1] * n

    def find(self, x: int) -> int:
        root = x
        while self.father[root] != root:
            root = self.father[root]
        while x != root:
            self.father[x], x = root, self.father[x]
        return root

    def isSameSet(self, x: int, y: int) -> bool:
        return self.find(x) == self.find(y)

    def union(self, x: int, y: int) -> bool:
        rootX, rootY = self.find(x), self.find(y)
        if rootX != rootY:
            xSize, ySize = self.size[rootX], self.size[rootY]
            big, small = (rootY, rootX) if xSize < ySize else (rootX, rootY)
            self.size[big] += self.size[small]
            self.father[small] = big
            self.count -= 1
        return False if rootX == rootY else True
